features/dd-sdk-android-rum/src/test/kotlin/com/datadog/android/rum/metric/ViewEndedMetricDispatcherTest.kt

```diff
@@ -0,0 +1,280 @@
+import fr.xgouchet.elmyr.annotation.FloatForgery
+import fr.xgouchet.elmyr.annotation.LongForgery
+import fr.xgouchet.elmyr.junit5.ForgeConfiguration
+import fr.xgouchet.elmyr.junit5.ForgeExtension
+import org.assertj.core.api.AssertionsForClassTypes.assertThat
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.extension.ExtendWith
+import org.junit.jupiter.api.extension.Extensions
+import org.mockito.Mock
+import org.mockito.junit.jupiter.MockitoExtension
+import org.mockito.junit.jupiter.MockitoSettings
+import org.mockito.kotlin.any
+import org.mockito.kotlin.argThat
+import org.mockito.kotlin.eq
+import org.mockito.kotlin.verify
+import org.mockito.kotlin.verifyNoMoreInteractions
+import org.mockito.quality.Strictness
+
+@Extensions(
+    ExtendWith(MockitoExtension::class),
+    ExtendWith(ForgeExtension::class)
+)
+@MockitoSettings(strictness = Strictness.LENIENT)
+@ForgeConfiguration(Configurator::class)
+internal class ViewEndedMetricDispatcherTest {
+
+    @Mock
+    lateinit var mockInternalLogger: InternalLogger
+
+    @FloatForgery(min = 0f, max = 1f)
+    private var fakeSampleRate: Float = 0f
+
+    @LongForgery(min = 0)
+    private var fakeDuration: Long = 0
+
+    @LongForgery(min = 0)
+    private var fakeLoadingTime: Long = 0
+
+    private lateinit var fakeViewType: RumViewType
+    private lateinit var fakeInvState: ViewInitializationMetricsState
+    private lateinit var fakeTnsState: ViewInitializationMetricsState
+
+    private lateinit var dispatcherUnderTest: ViewEndedMetricDispatcher
+
+    @BeforeEach
+    fun `set up`(forge: Forge) {
+        fakeViewType = forge.aValueFrom(RumViewType::class.java)
+        fakeTnsState = forge.aViewInitializationMetricsState(NoValueReason.TimeToNetworkSettle::class.java)
+        fakeInvState = forge.aViewInitializationMetricsState(NoValueReason.InteractionToNextView::class.java)
+
+        dispatcherUnderTest = ViewEndedMetricDispatcher(
+            viewType = fakeViewType,
+            internalLogger = mockInternalLogger,
+            samplingRate = fakeSampleRate
+        )
+    }
+
+    @Test
+    fun `M sendMetric only once W sendViewEnded`() {
+        // When
+        dispatcherUnderTest.sendViewEnded(fakeInvState, fakeTnsState)
+        // Then
+        verify(mockInternalLogger).logMetric(
+            messageBuilder = any(),
+            additionalProperties = any(),
+            samplingRate = eq(fakeSampleRate),
+            creationSampleRate = eq(null)
+        )
+
+        // When
+        dispatcherUnderTest.sendViewEnded(fakeInvState, fakeTnsState)
+
+        // Then
+        verify(mockInternalLogger).log(
+            eq(InternalLogger.Level.WARN),
+            target = eq(Target.TELEMETRY),
+            messageBuilder = any(),
+            throwable = eq(null),
+            onlyOnce = eq(false),
+            additionalProperties = eq(null)
+        )
+        verifyNoMoreInteractions(mockInternalLogger)
```

Comment:
minor: It is better to have a single `When` and single `Then` if possible. Here you can do both `dispatcherUnderTest.sendViewEnded` in `When` and have both assertions below, because they have different arguments to assert.